### PR TITLE
Use subcolon args to simplify optimizer options

### LINKED ISSUE #11873: optimizer command line flags cleanup
Split out of https://github.com/scala/bug/issues/11746#issuecomment-534015696

  - ~make `-opt:_` work~ done in https://github.com/scala/scala/pull/8454
  - `opt:inline` and `-opt:l:inline` could be confusing, we should probably rename `l:inline`
  - update https://www.lightbend.com/blog/scala-inliner-optimizer to mention the `-V` flags

### PR DESCRIPTION
**Pull request by som-snytt**
Optimizer is enabled by `-opt:inline:**`, as described in the [new overview](https://docs.scala-lang.org/overviews/compiler-options/optimizer.html). Warnings are enabled by `-Wopt` and verbose output by `-Vinline`.

Fixes scala/bug#11873

### CODE REVIEW

**Comment by som-snytt** [comment]
Must restore `l` (`ell`) for compatibility. Otherwise builds will break.

**Comment by som-snytt** [comment]
That was needlessly difficult and kludgy. Also I did not restore the `-opt-inline-from` and `-opt-warnings`. IIRC.

**Comment by som-snytt** [comment]
Wait, there's a 2.13.8 milestone?!

**Comment by som-snytt** [comment]
obviously, allowing `-opt:inline:**` breaks `-Wconf:cat=foo:s`

**Comment by som-snytt** [comment]
I thought I wasn't supposed to update the build, but I must have tossed a bad coin.
```
[error] -opt-inline-from is deprecated: use -opt:inline:**
[error] one error found
```
Edit: `setupPublishCore` needs old options
```
++ sbt -sbt-version 1.5.6 -Dsbt.override.build.repos=true -Dsbt.repository.config=/home/jenkins/workspace/scala-2.13.x-validate-main/scripts/sbt-repositories-config -warn 'setupPublishCore https://scala-ci.typesafe.com/artifactory/scala-pr-validation-snapshots/' publish
[error] 'all' is not a valid choice for '-opt'
[error] 'inline:scala/**' is not a valid choice for '-opt'
[error] 'all' is not a valid choice for '-opt'
[error] 'inline:scala/**' is not a valid choice for '-opt'
[error] 'all' is not a valid choice for '-opt'
[error] 'inline:scala/**' is not a valid choice for '-opt'
```
Edit: undeprecate the old options because no point in 2.13.

Also, some usages of `-opt:l:inline` in tests require only `-opt:inline`, others really do need `-opt:all` or a subset.

**Comment by lrytz** [comment]
Thank you for taking this on!

I see two issues with the current state:
  - `-opt:inline:pattern` only enables the inliner, no local optimizations. There is no situtaion I can think of for anyone to ever want this, local optimizations should always be on (except maybe us for testing the inliner). It's too easy to make that mistake - even this PR changes many `-opt:l:inline -opt-inline-from:**` to `-opt:inline:**`.
  - `-opt:inline` and `-opt:all` default to `-opt:inline:**`, which is unsafe. Since we're changing the flags, it's a chance to issue an error when specifying plain `-opt:inline` without a pattern.

Could we make `opt:inline:pattern` enalbe all optimizations, remove the `all` option, and require a `pattern`?

File: src/compiler/scala/tools/nsc/settings/ScalaSettings.scala [hunk 1]
```diff
@@ -387,13 +401,25 @@ trait ScalaSettings extends StandardScalaSettings with Warnings { _: MutableSett
...
-        |from expanding it to a list of files in the current directory.""".stripMargin))
+        |from expanding it to a list of files in the current directory.""".stripMargin
+
+  @deprecated("Deprecated alias", since="2.13.8")
+  val xoptInlineFrom = MultiStringSetting(
+    "-opt-inline-from",
+    "patterns",
+    "Patterns for classfile names from which to allow inlining, `help` for details.",
+    helpText = Some(inlineHelp))
+      //.withDeprecationMessage("use -opt:inline:**")
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:417]: would you deprecate it later?

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:417]: IIUC after re-starr? The problem was the build failing with the warning.

**Comment by som-snytt** [comment]
I will defer to your expert opinion on those two gotchas.

The only use case for fine-grained options is, "The optimizer broke my code, so I have to turn off a feature for this file."

I agree, it was a nuisance to fiddle with `-opt:all` in the tests. (I intended to enable only minimal options for testing, but didn't want to comprehend every test.)

On not defaulting to double star, it would be nice if there were a convenient default, such as, inline from classes with sources in this build. Then an option to inline from dependencies; from std lib; from platform lib. But maybe real programmers don't mind saying `com.example.**`. Oh wait, there is `<sources>`, why isn't that the default? Probably in sbt that only works for a clean build? and not for subprojects.

I will follow up with the tighter screws before I forget how it works.

**Comment by lrytz** [comment]
About defaulting to `<sources>`: it's already not a good idea today to enable the optimizer during development, but I wouldn't be surprised if it still happens. Defaulting to `<sources>` makes would make things even more random.

I think requiring people to think about inlining when enabling it is a good thing :)

**Comment by som-snytt** [comment]
Thanks lrytz, you have me singing the classic tune, "Puttin' on the Rytz". Probably I've made that pun before.

The new behavior is much better. One difference is that before, you needed `-opt:inline` or probably `-opt:l:inline` and `-opt-inline-from:**` to enable inlining, the way you need both keys turned simultaneously to launch the missiles.

Specifying only one option would silently not inline anything.

Now, `-opt:inline` requires the pattern, and `-opt-inline-from:**` is just an alias that works like `-opt:inline:**` so that by itself it enables.

File: build.sbt [hunk 2]
```diff
@@ -682,6 +682,7 @@ lazy val bench = project.in(file("test") / "benchmarks")
       if (benchmarkScalaVersion == "") Nil
       else "org.scala-lang" % "scala-compiler" % benchmarkScalaVersion :: Nil
     },
+    //scalacOptions ++= Seq("-feature", "-opt:all", "-opt:inline:scala/**", "-Wopt"),
```
Comment (lrytz) [inline build.sbt]:
```suggestion
    //scalacOptions ++= Seq("-feature", "-opt:inline:scala/**", "-Wopt"),
```

File: project/ScriptCommands.scala [hunk 3]
```diff
@@ -161,6 +161,7 @@ object ScriptCommands {
   }
 
   private[this] val enableOptimizer = Seq(
+    //ThisBuild / Compile / scalacOptions ++= Seq("-opt:all", "-opt:inline:scala/**")
```
Comment (lrytz) [inline project/ScriptCommands.scala]:
```suggestion
    //ThisBuild / Compile / scalacOptions ++= Seq("-opt:inline:scala/**")
```

**Comment by som-snytt** [comment]
I needed this the other day when I couldn't remember how to use `-opt`, so I used `-optimise`, with an `ess` in case it didn't accept `zed`. It was to check out `-Wperformance`.

File: src/compiler/scala/tools/nsc/settings/ScalaSettings.scala [hunk 4]
```diff
@@ -307,30 +307,31 @@ trait ScalaSettings extends StandardScalaSettings with Warnings { _: MutableSett
...
-      expandsTo = methodChoices)
-
-    private val inlineChoices = List(lMethod, inline)
-    val lInline = Choice("l:inline",
-      "Enable cross-method optimizations (note: inlining requires -opt-inline-from): " + inlineChoices.mkString("", ",", "."),
-      expandsTo = inlineChoices)
-
-    // "l:none" is excluded from wildcard expansion so that -opt:_ does not disable all settings
+      "local",
+      localOptimizations.mkString("Enable intra-method optimizations: ", ",", "."),
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/settings/ScalaSettings.scala]: `unreachableCode` is not in this list, should probably be included?

File: src/compiler/scala/tools/nsc/settings/ScalaSettings.scala [hunk 5]
```diff
@@ -307,30 +307,31 @@ trait ScalaSettings extends StandardScalaSettings with Warnings { _: MutableSett
...
+      expandsTo = defaultOptimizations ::: localOptimizations)
+
+    val inlineFrom = Choice(
+      "inline",
+      s"Inline method invocations from specified sites; enables all optimizations; see -Yopt-inline-heuristics.\n$inlineHelp",
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/settings/ScalaSettings.scala]:
```suggestion
      s"Inline method invocations, enables all optimizations. Requires specifying sites from where to inline, see below. See also -Yopt-inline-heuristics.\n\n$inlineHelp",
```

File: src/compiler/scala/tools/nsc/settings/ScalaSettings.scala [hunk 6]
```diff
@@ -363,17 +376,15 @@ trait ScalaSettings extends StandardScalaSettings with Warnings { _: MutableSett
...
 
-  val optInlineFrom = MultiStringSetting(
-    "-opt-inline-from",
-    "patterns",
-    "Patterns for classfile names from which to allow inlining, `help` for details.",
-    helpText = Some(
+  def optInlineFrom: List[String] = optChoices.inlineFrom.selections
+
+  def inlineHelp =
       """Patterns for classfile names from which the inliner is allowed to pull in code.
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/settings/ScalaSettings.scala]:
```suggestion
      """Inlining requires a list of patterns `-opt:inline:p1,p2` defining where code can be inlined from.
         |Patterns for classfile names from which the inliner is allowed to pull in code are:
```

File: src/compiler/scala/tools/nsc/settings/ScalaSettings.scala [hunk 7]
```diff
@@ -387,13 +398,22 @@ trait ScalaSettings extends StandardScalaSettings with Warnings { _: MutableSett
         |  <sources>      Classes defined in source files compiled in the current compilation, either
         |                 passed explicitly to the compiler or picked up from the `-sourcepath`
         |
-        |The setting accepts a list of patterns: `-opt-inline-from:p1,p2`. The setting can be passed
-        |multiple times, the list of patterns gets extended. A leading `!` marks a pattern excluding.
+        |The setting requires a list of patterns: `-opt:inline:p1,p2`. The setting can be passed
+        |multiple times, the list of patterns gets extended. A leading `!` marks a pattern as excluding.
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/settings/ScalaSettings.scala]:
```suggestion
        |The `-opt:inline:p1,p2` setting can be passed multiple times, the list of patterns gets extended.
        |A leading `!` marks a pattern as excluding.
```

**Review by lrytz — COMMENTED** [review]
LGTM, thank you very much! A few minor suggestions, if you agree I'm happy to push them, what ever works best.

File: src/compiler/scala/tools/nsc/settings/ScalaSettings.scala (hunk 4 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/settings/ScalaSettings.scala]: oh, in the help.

File: src/compiler/scala/tools/nsc/settings/ScalaSettings.scala (hunk 1 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:417]: The answer is no, I would forget.

File: src/compiler/scala/tools/nsc/settings/ScalaSettings.scala (hunk 5 above)
Comment (som-snytt) [reply src/compiler/scala/tools/nsc/settings/ScalaSettings.scala]: The reason this is unresolved is that I missed it.

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/settings/ScalaSettings.scala]: OK I pushed approximately that. I had already added a newline to the `inlineHelp`. It also follows `-opt-inline-from:help`.

**Comment by SethTisue** [comment]
@som-snytt do you want a second round of review from Lukas, or are you reasonably confident you've addressed his concerns and we should go ahead and merge?

**Comment by som-snytt** [comment]
yes, now it is quite gorgeous after following up his tweaks. It is tweakèd beauty.

**Comment by lrytz** [comment]
The old settings (`-opt:l:XX` and `-opt-inline-from:XX`) continue to work and without deprecation warning, but we should still update the PR description to explain what's changed and include it the release notes.

**Pull request merged**